### PR TITLE
Add support for new endpoints in DPLUS panels 

### PR DESCRIPTION
**Pull request by Poshy163**
With #75

### CODE REVIEW

**Comment by Poshy163** [comment]
Going to mark this as ready, but still going to test it within a dev container to HA to double check

**Comment by Poshy163** [comment]
Seems to work 😄 
image
image

**Review by nickw444 — COMMENTED** [review]
Do you think it's worth emulating any of this additional behaviour in the fake alarm server?

**Comment by Poshy163** [comment]
When connected to a [Ness-Bridge Ethernet Interface](https://ness.com.au/106-014), calls that have the CommandType of 0x66 are trying to be called, encapsulate these into their own event, and always return UNKNOWN_MEZZO_CONTROLLER_CALL


image

**Comment by Poshy163** [comment]
Cool. happy with this PR now, 

Both the mezzo bridge and a standard IP232 module now works successfully and shows the correct zones

**Comment by Poshy163** [comment]
Yeah, need to revert to the old way of doing it, seems like the mezzo bridge handles stuff in hex not binary. and there is no other real way to identify what the middle man is, other than just decoding its heartbeat calls, and determining off of that.

Unfortunately ness have no documentation on what any of the ASCII stuff on that is meant to be lol

**Comment by Poshy163** [comment]
image

**Comment by Poshy163** [comment]
alright @nickw444 its 100% good to go now, ness seem to be not wanting to reveal what is actually going on with the mezzo stuff, so will just keep it as it is, with a temporary "bypass" while still providing and logging what the event was that is received, just saves the logs getting spammed with ValueType errors  

image

**Review by nickw444 — CHANGES_REQUESTED** [review]
Sorry folks, struggling to find time to review this in depth in personal time. I will try find a slot on the weekend to take a deeper look.

Some initial directional feedback; I'm mostly uneasy about the global field on the class, ideally we an find a more pure way to model this logic? 

Finally I'm still of the opinion that I'd like to see this specific handling/behaviour modelled into the fake server. I would like to emphasise the importance of having this in the fake server because it is my only way to develop and verify changes to this library **as I do not own a ness alarm anymore** and for other contributors it would make it impossible/difficult to test features of different models without it.

**Comment by Poshy163** [comment]
after talking further with NESS regarding this, they seem pretty adamant that they do not want the module to "emulate" what standard IP232 module does (even tho it has the capability with some firmware tweaks). ill just go around and remove all the code regarding handling that, and just focus on the new DPLUS stuff

File: nessclient/event.py [hunk 1]
```diff
@@ -88,22 +91,25 @@ class EventType(Enum):
...
+            timestamp: Optional[datetime.datetime],
     ) -> None:
         super(SystemStatusEvent, self).__init__(address=address, timestamp=timestamp)
         self.type = type
         self.zone = zone
         self.area = area
 
     @classmethod
-    def decode(cls, packet: Packet) -> "SystemStatusEvent":
+    def decode(cls, packet: Packet, is_mezzo: bool = False) -> "SystemStatusEvent":
```
Comment (nickw444) [inline nessclient/event.py:107]: We should [remove](https://github.com/nickw444/nessclient/pull/76#issuecomment-3044658075) the `mezzo` handling right?

(and the handling below too)

File: examples/listening_for_events.py [hunk 2]
```diff
@@ -14,7 +14,7 @@ def on_zone_change(zone: int, triggered: bool):
 
 
 @client.on_state_change
-def on_state_change(state: ArmingState):
+def on_state_change(state: ArmingState, arming_mode):
```
Comment (nickw444) [inline examples/listening_for_events.py:17]: Please add a type to `arming_mode`.

File: nessclient/event.py [hunk 3]
```diff
@@ -32,11 +33,13 @@ def __repr__(self) -> str:
         return "<{} {}>".format(self.__class__.__name__, self.__dict__)
 
     @classmethod
-    def decode(cls, packet: Packet) -> "BaseEvent":
+    def decode(cls, packet: Packet) -> Union["BaseEvent", CommandType]:
```
Comment (nickw444) [inline nessclient/event.py:36]: Would it be possible to avoid this type-union in this instance? I'd consider this an antipattern - `BaseEvent::decode` should only ever return (child) instances of `BaseEvent`, not anything else.

Ideally we should avoid calling `BaseEvent::decode` entirely if we know `CommandType == CommandType.TEST_ASCII_STRING` in the parent caller.

```suggestion
    def decode(cls, packet: Packet) -> "BaseEvent":
```

File: nessclient/event.py [hunk 4]
```diff
@@ -88,22 +91,25 @@ class EventType(Enum):
         OUTPUT_OFF = 0x32
 
     def __init__(
-        self,
-        type: "SystemStatusEvent.EventType",
-        zone: int,
-        area: int,
-        address: Optional[int],
-        timestamp: Optional[datetime.datetime],
+            self,
```
Comment (nickw444) [inline nessclient/event.py:94]: Looks like some extraneous formatting quirks throughout, can you run `black` please.

File: nessclient/packet.py [hunk 5]
```diff
@@ -10,6 +10,14 @@
 class CommandType(Enum):
     SYSTEM_STATUS = 0x61
     USER_INTERFACE = 0x60
+    TEST_ASCII_STRING = 0X36
+
+    # This is a special command that is used to validate a "Test ASCII String" which may not be standard/documented but still contain 0x36.
+    @property
+    def requires_special_handling(self) -> bool:
```
Comment (nickw444) [inline nessclient/packet.py:17]: `required_special_handling` is too generic for a specific edge case, it doesn't provide any meaning, e.g. why the special handling is required / what it is for. Instead, the code could be better self-documenting by making the test inline, including reasoning for the special handling.

File: nessclient/packet.py [hunk 6]
```diff
@@ -119,6 +127,9 @@ def decode(cls, _data: str) -> "Packet":
         if has_timestamp(start):
             timestamp = decode_timestamp(data.take_bytes(6))
 
+        if command.requires_special_handling:
```
Comment (nickw444) [inline nessclient/packet.py:130]: e.g. instead `if command == CommandType.TEST_ASCII_STRING:`

We have a similar inline comparison on L141.

File: nessclient/packet.py [hunk 7]
```diff
@@ -119,6 +127,9 @@ def decode(cls, _data: str) -> "Packet":
...
+            return Packet(address, seq, command, msg_data, timestamp, True)
```
Comment (nickw444) [inline nessclient/packet.py:131]: I couldn't really understand why this `return` is different to the one on `L139`. Can we fold this into the same return statement? It's unclear what is truly "special" about this branch.

**Review by nickw444 — CHANGES_REQUESTED** [review]
Sorry for the delay, have done another pass.

**Comment by nickw444** [comment]
Hi @Poshy163, I've worked on https://github.com/nickw444/nessclient/pull/105 last night - At a basis the changes in this PR are insufficient for handling 32 zones, a few key issues I realised when building upon the implementation:

* `ZONE_17_32_INPUT_UNSEALED` events are unhandled (this is how the panel informs the client that a zone 17-32 state is changed)
* `ZONE_17` to `ZONE_32` in the `ZoneUpdate.Zone` enum have duplicated values (because the protocol can only encode values up to zone 16. Which explains why there is an introduction of `ZONE_17_32_*` events. 

You will notice in #105 that I ensure that `event.py` event classes intent to model/map closely to the protocol spec, that has meant creating a new `ZoneUpdate_17_32` class, then using `alarm.py` to be the abstraction on top of these, to map into a single bank of zones.

**Comment by Poshy163** [comment]
Makes complete sense, ill close this PR and will focus on improving yours, as this one has blown way out of scope. 

Ive been working with joost (via the discord), on setting up a config_flow for the integration, but ill wait on these PR's to flow through to pypi before i proceed any further (not that it should make a difference, but would be nice to test with what will be in the release)

**Comment by nickw444** [comment]
@Poshy163 the new `send_command_and_wait` method will probably be useful for some of the config flow stuff too for probing the alarm and receiving responses inline. #107
